### PR TITLE
Adding STORM32_V1_31_RC variant

### PR DESCRIPTION
**Pull request by owennewo**
**Summary**

Adding STORM32_V1_31_RC variantI

I've placed this in the ESC_board menu - technically it is a 3axis gimbal controller (not an ESC) but it can be used as an ESC for smaller motors.

This is a very popular board designed by Olliw (I have no affiliation).
http://www.olliw.eu/storm32bgc-wiki/Main_Page

**Validation**
I've tested this using SimpleFOC library and have most things working
 - SerialUSB, Serial3, I2C1, I2C2
 - Can spin 3 motors 

Here is an example sketch of it in use:
https://gist.github.com/owennewo/68c3de58d3e14b702552f0c149de41b0

### CODE REVIEW

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 1]
```diff
@@ -0,0 +1,300 @@
+
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:1]: Missing Header comments
```suggestion
/*
 *******************************************************************************
 * Copyright (c) 2020-2021, STMicroelectronics
 * All rights reserved.
 *
 * This software component is licensed by ST under BSD 3-Clause license,
 * the "License"; You may not use this file except in compliance with the
 * License. You may obtain a copy of the License at:
 *                        opensource.org/licenses/BSD-3-Clause
 *
 *******************************************************************************
 */
/*
 * Automatically generated from STM32F103R(C-D-E)Tx.xml
 * CubeMX DB release 6.0.21
 */
```

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  {NC,         NP,     0}
+};
+#endif
+
+#ifdef HAL_UART_MODULE_ENABLED
+WEAK const PinMap PinMap_UART_RTS[] = {
+  {PA_1,       USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART2_DISABLE)},
+  {PA_12,      USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PB_14,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PB_14_ALT1, USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:193]: To save space
```suggestion
  // {PA_1,       USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART2_DISABLE)},
  // {PA_12,      USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  // {PB_14,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  // {PB_14_ALT1, USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  {NC,         NP,     0}
+};
+#endif
+
+#ifdef HAL_UART_MODULE_ENABLED
+WEAK const PinMap PinMap_UART_CTS[] = {
+  {PA_0,       USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PA_11,      USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PB_13,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_DISABLE)},
+  {PB_13_ALT1, USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:203]:
```suggestion
  // {PA_0,       USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  // {PA_11,      USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  // {PB_13,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_DISABLE)},
  // {PB_13_ALT1, USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  {PB_13,     SPI2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {NC,        NP,   0}
+};
+#endif
+
+#ifdef HAL_SPI_MODULE_ENABLED
+WEAK const PinMap PinMap_SPI_SSEL[] = {
+  {PA_4,       SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_DISABLE)},
+  {PA_15,      SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
+  {PA_15_ALT1, SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:244]:
```suggestion
  // {PA_4,       SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_DISABLE)},
  // {PA_15,      SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
  // {PA_15_ALT1, SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
```

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+#endif
+
+#ifdef HAL_UART_MODULE_ENABLED
+WEAK const PinMap PinMap_UART_RX[] = {
+  // {PA_3,       USART2, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART2_DISABLE)},
+  {PA_10,      USART1, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART1_DISABLE)},
+  // {PB_7,       USART1, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART1_ENABLE)},
+  // {PB_11,      USART3, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+  // {PC_11,      UART4,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+  {PC_11, USART3, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:182]: Note that other way would be to use  `PC11_ALT1` in the variant header
`#define USART3_RX       PC11_ALT1`

```suggestion
  // {PC_11_ALT1, UART4,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
  {PC_11,      USART3, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```

File: variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+//*** UART ***
+
+#ifdef HAL_UART_MODULE_ENABLED
+WEAK const PinMap PinMap_UART_TX[] = {
+  // {PA_2,       USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART2_DISABLE)},
+  {PA_9,       USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART1_DISABLE)},
+  // {PB_6,       USART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART1_ENABLE)},
+  // {PB_10,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  // {PC_10,      UART4,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PC_10, USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/PeripheralPins_STORM32_V1_31_RC.c:169]: Note that other way would be to use PC10_ALT1 in the variant header
`#define USART3_TX       PC10_ALT1`

```suggestion
  // {PC_10_ALT1, UART4,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  {PC_10,      USART3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Hi @owennewo 
Thanks for your contribution.
Nice first PR moreover you are the first contributor with the new format.
Hope it was not so hard, wiki will be updated soon.

In addition to my comments could you also update the README.md to reference it in the ESC section.
Moreover it is not very clear for me which board STorM32 board you add?
Is it this one ? http://www.olliw.eu/storm32bgc-wiki/STorM32_Boards#STorM32-BGC_v1.3
In that case why `RC` instead of `BGC` ?

Tthank you in advance.

**Comment by owennewo** [comment]
I've repushed with a README.

Yes that link is correct.  RC refers to STM32F103*RC* which is by far the most common flash size version (and is the one I have), but as this board is open source, people have made RB, RE and RG versions which I've not tested and don't really know whether they'll work.  If you'd prefer me to remove the RC from the name, so that others can later add support for different flash sizes then I can do that.  I can also add BGC (which I think stands for Brushless Gimbal Controller) if you like.

**Comment by fpistm** [comment]
Thanks for the README.
About the different flash size, in fact it can be managed easily by adding correct boards entry in the boards.txt and renaming the `variant_STORM32_V1_31_RC.*`  and `peripheralPins_STORM32_V1_31_RC.c ` to 
`variant_STORM32_V1_31_Rx.*` and `peripheralPins_STORM32_V1_31_Rx.c `

Then adding in the boards.txt:
+# STORM32_V1_RC board
+ESC_board.menu.pnum.STORM32_V1_31_RC=STorM32 BGC v1.31 RC
+ESC_board.menu.pnum.STORM32_V1_31_RC.upload.maximum_size=262144
+ESC_board.menu.pnum.STORM32_V1_31_RC.upload.maximum_data_size=49152
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.mcu=cortex-m3
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.board=STORM32_V1_31_RC
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.series=STM32F1xx
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.product_line=STM32F103xE
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.cmsis_lib_gcc=arm_cortexM3l_math
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.variant=STM32F1xx/F103R(C-D-E)T
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.variant_h=variant_STORM32_V1_31_Rx.h

+# STORM32_V1_RD board
+ESC_board.menu.pnum.STORM32_V1_31_RD=STorM32 BGC v1.31 RD
+ESC_board.menu.pnum.STORM32_V1_31_RD.upload.maximum_size=393216
+ESC_board.menu.pnum.STORM32_V1_31_RD.upload.maximum_data_size=65536
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.mcu=cortex-m3
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.board=STORM32_V1_31_RD
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.series=STM32F1xx
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.product_line=STM32F103xE
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.cmsis_lib_gcc=arm_cortexM3l_math
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.variant=STM32F1xx/F103R(C-D-E)T
+ESC_board.menu.pnum.STORM32_V1_31_RD.build.variant_h=variant_STORM32_V1_31_Rx.h

+# STORM32_V1_RE board
+ESC_board.menu.pnum.STORM32_V1_31_RE=STorM32 BGC v1.31 RE
+ESC_board.menu.pnum.STORM32_V1_31_RE.upload.maximum_size=524288
+ESC_board.menu.pnum.STORM32_V1_31_RE.upload.maximum_data_size=65536
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.mcu=cortex-m3
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.board=STORM32_V1_31_RE
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.series=STM32F1xx
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.product_line=STM32F103xE
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.cmsis_lib_gcc=arm_cortexM3l_math
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.variant=STM32F1xx/F103R(C-D-E)T
+ESC_board.menu.pnum.STORM32_V1_31_RE.build.variant_h=variant_STORM32_V1_31_Rx.h

And in each files adding the correct switch:
`#if defined(ARDUINO_STORM32_V1_31_RC) || defined(ARDUINO_STORM32_V1_31_RD) || defined(ARDUINO_STORM32_V1_31_RE)`

In fact the linker script is preprocessed so the correct flash and RAM size will be used:
https://github.com/stm32duino/Arduino_Core_STM32/blob/0a3fa54b4a645eba9f9e5db9165a85c3bfc6d66e/variants/STM32F1xx/F103R(C-D-E)T/ldscript.ld#L36-L41

Note that you didn't fix my other review comments.

File: boards.txt [hunk 7]
```diff
@@ -4931,6 +4931,20 @@ ESC_board.menu.pnum.WRAITH32_V1.build.variant=STM32F0xx/F051K(6-8)U
...
+ESC_board.menu.pnum.STORM32_V1_31_RC.upload.maximum_data_size=49152
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.mcu=cortex-m3
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.board=STORM32_V1_31_RC
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.series=STM32F1xx
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.product_line=STM32F103xE
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.cmsis_lib_gcc=arm_cortexM3l_math
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.variant=STM32F1xx/F103R(C-D-E)T
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.variant_h=variant_{build.board}.h
+ESC_board.menu.pnum.STORM32_V1_31_RC.build.extra_flags=-D{build.product_line} {build.enable_usb} {build.xSerial}
```
[the remark this answers is not in the document]
Comment (owennewo) [reply boards.txt:4946]: Aah! this is why my batch commit (of your suggestions) wouldn't work.  Github us saying
`Suggestion cannot be identical to original text. `

You meant to delete the two lines?

**Comment by owennewo** [comment]
I'll retest after accepting your changes.  The Serial3/SerialUSB took me a while to get working automatically so this is the area I'm not confident in - perhaps something else is missing in this area.

Regaring RC/RB/RE/RG versions of this board - I think I confused myself and you

RC - exists
RE - only exists in a later variant (3.x which is quite different)
RB and RG are mentioned in this schematic https://community.simplefoc.com/uploads/default/original/1X/6f1cdb4a8d6bfdd6d405aa340ca2571f5be42738.jpeg but I've not seen them for sale.  I don't think they exist!
TLDR: I think we only need single flash size RC

File: boards.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (fpistm) [reply boards.txt:4946]: yes

File: README.md [hunk 8]
```diff
@@ -432,6 +432,7 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
 | Status | Device(s) | Name | Release | Notes |
 | :----: | :-------: | ---- | :-----: | :---- |
 | :green_heart:  | STM32F051K6U | [Wraith V1 ESC](https://www.readytoflyquads.com/35a-wraith32-v11-32bit-blheli-esc) | *1.8.0* |  |
+| :green_heart:  | STM32F103RCT | [STorM32 V1.31 RC](http://www.olliw.eu/storm32bgc-wiki/STorM32_Boards#STorM32-BGC_v1.3) | *2.0.0* |  |
```
Comment (fpistm) [inline README.md:435]:
```suggestion
| :yellow_heart:  | STM32F103RCT | [STorM32 V1.31 RC](http://www.olliw.eu/storm32bgc-wiki/STorM32_Boards#STorM32-BGC_v1.3) | **2.1.0** |  |
```

**Comment by fpistm** [comment]
OK so we can kept only one variant.


>  The Serial3/SerialUSB took me a while to get working automatically so this is the area I'm not confident in - perhaps something else is missing in this area.

I'm confident 😉

File: variants/STM32F1xx/F103R(C-D-E)T/variant_STORM32_V1_31_RC.h [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  #define PIN_SERIAL_TX         USART3_TX
+#endif
+
+// Extra HAL modules
+#if !defined(HAL_DAC_MODULE_DISABLED)
+  #define HAL_DAC_MODULE_ENABLED
+#endif
+#if !defined(HAL_SD_MODULE_DISABLED)
+  #define HAL_SD_MODULE_ENABLED
+#endif
```
Comment (fpistm) [inline variants/STM32F1xx/F103R(C-D-E)T/variant_STORM32_V1_31_RC.h:129]: It seems there is no SD card slot on the board and SDIO pins are used for other stuff. So I think the define could be safely removed.
```suggestion
```

**Review by fpistm — COMMENTED** [review]
PR LGTM. Only one question about the `HAL_SD_MODULE_ENABLED` .

**Comment by owennewo** [comment]
> OK so we can kept only one variant.
> 
> > The Serial3/SerialUSB took me a while to get working automatically so this is the area I'm not confident in - perhaps something else is missing in this area.
> 
> I'm confident

I found the issue - ESC_board.menu was missing `CDC (generic 'Serial' supersede U(S)ART)` (I was using `CDC (No generic Serial)` )

Hopefully this is ready to merge

File: boards.txt [hunk 10]
```diff
@@ -5857,6 +5869,15 @@ Midatronics.menu.xusb.HS.build.usb_speed=-DUSE_USB_HS
...
 Midatronics.menu.xusb.HSFS.build.usb_speed=-DUSE_USB_HS -DUSE_USB_HS_IN_FS
 
+ESC_board.menu.usb.none=None
+ESC_board.menu.usb.CDCgen=CDC (generic 'Serial' supersede U(S)ART)
+ESC_board.menu.usb.CDCgen.build.enable_usb={build.usb_flags} -DUSBD_USE_CDC
+ESC_board.menu.usb.CDC=CDC (no generic 'Serial')
+ESC_board.menu.usb.CDC.build.enable_usb={build.usb_flags} -DUSBD_USE_CDC -DDISABLE_GENERIC_SERIALUSB
+ESC_board.menu.xusb.FS=Low/Full Speed
+ESC_board.menu.xusb.HS.build.usb_speed=-DUSE_USB_HS
+ESC_board.menu.xusb.HSFS.build.usb_speed=-DUSE_USB_HS -DUSE_USB_HS_IN_FS
```
Comment (fpistm) [inline boards.txt:5879]: Add missing entry menu. Not needed for this board specifivally but generic for ESC boards menu entry.

```suggestion
ESC_board.menu.xusb.HS=High Speed
ESC_board.menu.xusb.HS.build.usb_speed=-DUSE_USB_HS
ESC_board.menu.xusb.HSFS=High Speed in Full Speed mode
ESC_board.menu.xusb.HSFS.build.usb_speed=-DUSE_USB_HS -DUSE_USB_HS_IN_FS
```

**Review by fpistm — APPROVED** [review]

**Comment by fpistm** [comment]
> I found the issue - ESC_board.menu was missing `CDC (generic 'Serial' supersede U(S)ART)` (I was using `CDC (No generic Serial)` )

Nice catch I missed that. 🙄 

> 
> Hopefully this is ready to merge

Yes, I've made a small update of USB menu entry.
